File: stream/search/schema.go
package main

import (
	_ "embed"
	"encoding/json"
	"log"
	"net/http"
	"strings"
)

// The canonical media vocabulary, embedded from the one file the web client
// also reads. Two hand-kept lists is exactly how `kind=movies` came to match
// nothing while the same request spelled `groups=movies` worked: the mapping
// existed on one side only. Embedding makes drift a build-time impossibility
// rather than a thing to remember.
//
//go:embed schema.json
var schemaJSON []byte

type mediaDomain struct {
	Label   string   `json:"label"`
	Verb    string   `json:"verb"`
	Types   []string `json:"types"`
	Aliases []string `json:"aliases"`
}

type mediaSchema struct {
	Version        int                    `json:"version"`
	Domains        map[string]mediaDomain `json:"domains"`
	Roles          map[string]string      `json:"roles"`
	Capabilities   []string               `json:"capabilities"`
	Health         []string               `json:"health"`
	ChannelSources []string               `json:"channelSources"`
}

var (
	schema mediaSchema
	// alias -> canonical domain id. Built once; every lookup is a map hit.
	domainByAlias = map[string]string{}
	// type -> canonical domain id, so a caller may name either.
	domainByType = map[string]string{}
)

func init() {
	if err := json.Unmarshal(schemaJSON, &schema); err != nil {
		// Unrecoverable and always a build-time mistake: the binary carries a
		// malformed vocabulary and nothing downstream can be trusted.
		log.Fatalf("media schema is not valid JSON: %v", err)
	}
	for id, d := range schema.Domains {
		domainByAlias[id] = id
		for _, a := range d.Aliases {
			domainByAlias[normaliseToken(a)] = id
		}
		for _, t := range d.Types {
			domainByType[normaliseToken(t)] = id
		}
	}
}

// normaliseToken folds the spellings that reach an API from a URL, a TV remote
// or a hand-written config: case, spacing and separators.
func normaliseToken(s string) string {
	s = strings.ToLower(strings.TrimSpace(s))
	s = strings.NewReplacer("-", "", "_", "", " ", "").Replace(s)
	return s
}

// canonicalDomain resolves any name a caller might use onto a domain id.
//
// Returns "" for anything unrecognised, which callers must treat as "no
// filter" rather than "match nothing". A filter nothing can satisfy is
// indistinguishable from a broken server; showing everything is at least a
// visible, arguable answer.
func canonicalDomain(s string) string {
	if s == "" {
		return ""
	}
	t := normaliseToken(s)
	if id, ok := domainByAlias[t]; ok {
		return id
	}
	return domainByType[t] // "" when unknown
}

// sameDomain compares two names through the canonical form.
//
// Both sides are normalised on purpose. A cached card may still carry "audio"
// while a new client sends "music"; comparing raw strings would silently drop
// every one of those cards, which is the original bug wearing a different hat.
func sameDomain(a, b string) bool {
	ca, cb := canonicalDomain(a), canonicalDomain(b)
	if ca == "" || cb == "" {
		return false
	}
	return ca == cb
}

// handleSchema serves the vocabulary to clients that cannot embed it at build
// time -- Roku, Tizen, anything on a TV. Those clients are the ones that were
// bitten last time, precisely because their copy of the mapping was written by
// hand and then drifted.
func (s *server) handleSchema(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	// Safe to cache: the vocabulary changes with a deploy, not with a request.
	w.Header().Set("Cache-Control", "public, max-age=3600")
	_, _ = w.Write(schemaJSON)
}

File: stream/web/src/schema.js
/**
 * The canonical media vocabulary, imported from the same file the Go server
 * embeds.
 *
 * This is a relative import into ../../search/ rather than a copy, and that is
 * the whole point. The previous arrangement kept the mapping in two places: the
 * browser sent `groups=movies`, the server compared against `video`, and every
 * card was filtered out while the response still reported hundreds of results.
 * Nothing errored, so it survived until someone counted the rows.
 *
 * Clients that cannot import at build time -- Roku, Tizen -- fetch the same
 * bytes from /api/schema.
 */

import schema from '../../search/schema.json' with { type: 'json' };

export const SCHEMA_VERSION = schema.version;
export const DOMAINS = schema.domains;
export const ROLES = schema.roles;
export const CAPABILITIES = schema.capabilities;
export const HEALTH_STATES = schema.health;
export const CHANNEL_SOURCES = schema.channelSources;

/** Fold the spellings that arrive from a URL, a remote, or a config file. */
function token(s) {
  return String(s ?? '').trim().toLowerCase().replace(/[-_\s]/g, '');
}

const byAlias = new Map();
const byType = new Map();
for (const [id, d] of Object.entries(DOMAINS)) {
  byAlias.set(token(id), id);
  for (const a of d.aliases || []) byAlias.set(token(a), id);
  for (const t of d.types || []) byType.set(token(t), id);
}

/**
 * Resolve any name onto a canonical domain id.
 *
 * Returns '' for anything unknown, and callers must read that as "no filter"
 * rather than "match nothing". A filter nothing can satisfy looks identical to
 * a broken server; showing everything is at least an answer someone can argue
 * with.
 */
export function canonicalDomain(name) {
  const t = token(name);
  if (!t) return '';
  return byAlias.get(t) ?? byType.get(t) ?? '';
}

/**
 * Compare two names through their canonical forms.
 *
 * Both sides are normalised deliberately: a card cached before this vocabulary
 * existed still says 'audio' where a new client says 'music'.
 */
export function sameDomain(a, b) {
  const ca = canonicalDomain(a);
  const cb = canonicalDomain(b);
  return ca !== '' && ca === cb;
}

/** Which domain a media type belongs to: 'episode' -> 'video'. */
export function domainOfType(type) {
  return byType.get(token(type)) ?? '';
}

/** The verb for a domain -- watch, listen, read, play, view. */
export function verbFor(name) {
  const id = canonicalDomain(name);
  return id ? DOMAINS[id].verb : '';
}

/** Human label for a domain, for tabs and headings. */
export function labelFor(name) {
  const id = canonicalDomain(name);
  return id ? DOMAINS[id].label : '';
}

/** Every canonical domain id, in a stable order for building UI. */
export function allDomains() {
  return Object.keys(DOMAINS);
}

/**
 * Does this provider advertise the role needed for an action?
 *
 * Callers must ask rather than assume. Radarr has no guide and never will;
 * asking it for one is how a universal front door turns into an if-type tree.
 */
export function hasRole(provider, role) {
  return Array.isArray(provider?.roles) && provider.roles.includes(role);
}

export function hasCapability(provider, capability) {
  return Array.isArray(provider?.capabilities) && provider.capabilities.includes(capability);
}
